python/ml4ir/base/model/calibration/temperature_scaling.py

```diff
@@ -0,0 +1,328 @@
+    in a .zip file.
+
+    Parameters
+    ----------
+        model :  tf.keras.Model
+                 Model object to be used for temperature scaling
+        scorer:  ScorerBase object
+                 ScorerBase object of the RelevanceModel
+        dataset: RelevanceDataset
+                 RelevanceDataset object to be used for training and evaluating temperature scaling
+        logger : Logger
+                 Logger object to log events
+        logs_dir_local: str
+                 path to save the TS scores
+        temperature_init: float
+                temperature initial value
+        file_io:FileIO
+                file I/O handler objects for reading and writing data
+
+    Returns
+    -------
+        `Union[np.ndarray, Tuple[np.ndarray, ...]]`
+        optimizer output containing temperature value learned during temperature scaling
+    """
+    # get an intermediate model with logits as output --> MUST NOT BE AN ACTIVATION (e.g.
+    # SOFTMAX) LAYER
+    intermediate_model = get_intermediate_model(model, scorer)
+
+    start = tf.Variable(initial_value=[temperature_init], shape=(1), dtype=tf.float32, name="temp",
+                        trainable=True)
+
+    def nll_loss_fn(temperature):
+        """Computes Negative Log Likelihood loss by applying temperature scaling.
+        Parameters:
+            temperature: TF.Tensor,  a  parameter of size=(1) to be trained
+        Returns:
+            NLL loss value (averaged on total examples)
+        """
+        logits_tensor = tf.constant(logits_nps, name='logits_tensor', dtype=tf.float32)
+        labels_tensor = tf.constant(labels_nps, name='labels_tensor', dtype=tf.int32)
+        logits_w_temp = tf.divide(logits_tensor, temperature)
+
+        return tf.reduce_mean(
+            tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels_tensor,
+                                                           logits=logits_w_temp))
+
+    @tf.function
+    def nll_with_lbfgs():
+        """Returns optimizer function. Inspired by
+         https://github.com/gpleiss/temperature_scaling/blob/master/temperature_scaling.py#L60"""
+        return tfp.optimizer.lbfgs_minimize(make_val_and_grad_fn(nll_loss_fn),
+                                            initial_position=start)
+
+    # evaluation on validation set before temperature scaling
+    logits_nps, labels_nps = get_logits_labels(intermediate_model, dataset.validation)
+    original_acc_op, original_nll_loss_op, _ = eval_relevance_model(scorer, logits_nps,
+                                                                    labels_nps)
+
+    # perform temperature scaling
+    results = run_optimizer(nll_with_lbfgs, logger)
+
+    # evaluation on validation set with temperature scaling
+    temper = tf.constant(results.position)
+    acc_op, _, _ = eval_relevance_model(scorer, logits_nps, labels_nps, temperature=temper)
+
+    logger.info("=" * 50)
+    logger.info(f'temperature value : {results.position}')
```

Comment:
@marjanhs Can you add the temperature value to the `experiment_tracking_dict` https://github.com/salesforce/ml4ir/blob/master/python/ml4ir/base/pipeline.py#L390 ?

Reply:
added